### ISSUE TITLE
Bug: await-thenable doesn't handle unions cleanly in for-await [8.8.0 regression]

### ISSUE THREAD

**Issue by JavaScriptBach** [typescript-eslint/typescript-eslint#10080]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&showAST=es&fileType=.tsx&code=MYewdgzgLgBAhhAnmYBJKBTATgLhgQSRXWzgCMAbDAHmiwEswBzAPhgB8YStyraoGzNgF4YAbQC6AbgBQAMxBZ4Adzj1YAClCRYEGCDnwiaTFgCUMAN4wAvjIwAPAA6LY1m0A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOiiJo0APbRI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const asyncIter: AsyncIterable<string> | Iterable<string> = [];

for await (const s of asyncIter) { }

export { }
```

### ESLint Config

```javascript
See playground
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

If I call `for await` on an iterator that is typed as `AsyncIterable<string> | Iterable<string>`, I expect one of the following to happen:

1. No lint error, because the code is arguably correct (`for await` works on both sync and async iterables)
2. Lint error but no autofix, because the code is arguably ugly (taking a union of sync and async iterables leads to unnecessary promises being created, which is the whole point of this rule), but it's not safe to remove the `await` because that would introduce a code bug.

### Actual Result

Lint error, with an unsafe autofix suggesting to remove the `await`.

**Comment by kirkwaiblinger** [comment]
@JavaScriptBach could you provide the playground link or examples please?

**Comment by kirkwaiblinger** [comment]
Hi @JavaScriptBach!

Could you expand a bit on what a a concrete use case looks like when this comes up, and why it would make sense in general to use `for await` there?

Btw for other readers, here is the sample code
```ts
const asyncIter: AsyncIterable<string> | Iterable<string> = [];
for await (const s of asyncIter) { }
export { }
```

**Comment by kirkwaiblinger** [comment]
Note also that there is a _suggestion_, but no _autofix_, since the fix would not be safe in general.

**Comment by JavaScriptBach** [comment]
Ah sorry you're right, it is a suggestion, not an autofix.

Closing; it looks like everything here is working as intended.

My use case is that I have some library code that intentionally takes in a parameter typed `AsyncIterable<T> | Iterable<T>` and uses `for await` to handle both cases without repeating code. But I can disable the lint rule in that case as mentioned in the docs.

**Comment by kirkwaiblinger** [comment]
Well, the check for async iterables was just added this morning 🙂 https://github.com/typescript-eslint/typescript-eslint/pull/10008... so it may be "working as intended", but we're definitely listening closely at this time for feedback on whether the design is good in the first place or not! 

I'd definitely invite you to continue the discussion here if you would like to make a case for an improvement to the behavior around `for await` checks.

**Comment by JavaScriptBach** [comment]
This rule has a high batting average on my company's codebase, so overall I'm in favor of it.

It gives a false positive when operating on union types that may need `for await`, so perhaps you could consider make the rule not error in those cases? But I'm not familiar enough with type theory to know if that's an easy thing to detect or not.

**Comment by kirkwaiblinger** [comment]
> This rule has a high batting average on my company's codebase, so overall I'm in favor of it.

❤️ 

> It gives a false positive when operating on union types that may need `for await`, so perhaps you could consider make the rule not error in those cases? But I'm not familiar enough with type theory to know if that's an easy thing to detect or not.

This would definitely be simple to implement. I'm just looking to understand what the justification to do so is.

**Comment by bradzacher** [comment]
I think that this should be considered a bug personally because the rule does not report on a similar union case for non-iterables:

[playground](https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=MYewdgzgLgBAhhAnmYBJKBTATgLhgQSRXWzgCMAbDAHmiwEswBzAPhgB8YStyraoGzNgF4YAbQC6AbgBQAMxBZ4Adzj1YAClCRYEGCDnwiaTFgCUMAN4wAvjPsATDMApwsGGNugwAtnERkGAAKWCA%2B9BAYeCFhETRgAK4%2BgVhsnInJ2LJwquq%2B-oEx4ZGyMhgAHgAOirDWNkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOiiJo0APbRI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by kirkwaiblinger** [comment]
> I think that this should be considered a bug personally because the rule does not report on a similar union case for non-iterables

That is true, however, `await nonThenable` is much more of a harmless noop than `for await(const x of syncIterable) {}`, which can have undesirable effects on error-handling.

**Comment by kirkwaiblinger** [comment]
in other words, it's quite intentional that the rule flags on `for await` to on sync iterables, even as a convenient way of unwrapping promises, since it has meaningfully different control flow from async iteration.

```ts
declare const arrayOfPromises: Array<Promise<number>>;
for await (const x of arrayOfPromises) { // bad
    x satisfies number;
}
```

But perhaps being somewhat lax with the union case is a worthwhile devX concession for useful real-world scenarios. That's the angle by which I'm the most likely to be persuaded.

**Comment by kirkwaiblinger** [comment]
(see also https://github.com/typescript-eslint/typescript-eslint/issues/8858#issuecomment-2148977352)

**Comment by bradzacher** [comment]
Yeah I would be looking at it from the perspective that if you're attempting to iterate a `AsyncIterable<string> | Iterable<string>` then there is no option for you to separate out those types, right?

You could probably fashion some code that allows you to split things into two loops - one that's async and one that's not. But my question would be - why would you do that when that would require you to duplicate code.

Its the same reason we don't force people to do this for the regular promise union case:
```ts
const result = (value instanceof Promise) ? await value : value;
```

Sure this code is "more correct" but it's also not any better or any clearer.

So yeah I'd argue that from a DevX perspective flagging this is technically correct, but realistically it's fine to await this specific union case.

**Comment by kirkwaiblinger** [comment]
Yeah, I think I was radicalized by a couple of proposals ([this](https://es.discourse.group/t/syncandasynciterator/554) and [this](https://es.discourse.group/t/take-2-generator-prototype-symbol-mixediterator/1308)) I read a few years ago to believe that it was specifically bad to mix async and sync iterators, but that's probably not terribly relevant here.

The main difference I see between the `await maybePromise` and `for await (const {} of maybeAsyncIterable)` cases is that the `for await` has meaningful differences to do with control flow.

So, whereas the difference between the following two lines
```ts
declare const value: Promise<string> | string;
const version1 = (value instanceof Promise) ? await value : value;
const version2 = await value;
```
is basically only observable via the incurring of an extra microtask tick, the `for await` vs `for` causes more standard control flow questions....
([TS Playground](https://www.typescriptlang.org/play/?target=99#code/GYVwdgxgLglg9mABAJwKYGc4BsBuqCCwUqyAFALboBciYI5ARiQJSIDeAUIiqlCMkjCoA7ogAKyOORjpUAHhxwYAEwB8pNJlypEAXlXtEsqABUY5VHBBRSpVvp5a8dgDSJKrAL7MA3B08cHKCQsAiIAFSIAJ4wqFjKAIwATADMAMpRkHbsXNzRsfGICT75ccqISSUxZYgpfgEcAIbomRCIwdDwSJHV8ckp+C1ZrGyIudyNwo0wUI7YeITEZAkADCu+46WFxWN5iJPTs5rzBEQkpKvrfnu95ZW7eQczc9qL55cbNwXldf6BHaFultlBIpDJUP0MsMcl8asdXmdlmtmAA6KAAC1QYFs9gMCV8D24txeC0RF2RaMx2OyDiSBM2xPhpKW5PWlKxOL0BhSGwazVa7XAnTCAAdkDAwFAAGqNLAgDCDVpyEzqGYkRoMLCoGgASSWGq1yoMAB9EIrIHr1Zr5CqRrkIAh0LMnY1kKZzDpdIgRSRgHBkORGpBUCiwHBhHZrogHWAtCGsHAAOakADkLrdEsTiCgcEQauQjWIiFI-Mg7SwjUUyGYKc+7X9+ymz1IMadiBwsvliDgwDz+utdr2rew8aTpA7ctQdYC3FbsyxyjMFj03t9-sDwdD4cj9sdI5RCeTAANlAgdPnC5ni6W2sAK1XUYgAMJSEVa4jlRqIxAAEjYC6XHQAFojCgV13QsTx3HQI8Nj+QIb0FEIum9cVJRlSd0AAWRgAAPVBlCNUgL2tXV+0NFVEFNc0IEtAtrSNQdo0dZ0wLdQCVx9ZA-QDIMIBDMMIzrYctQPMc0zY2AwCzHM+3VIsKDwgjy0rf0azrGBe1IDJGGwFEbzowsGwlOT6K1ZhOD2HjG0OYs53bTsdB7UyDSnGE9mY2N90PcdHLrbgZ0QKC4lkdzuGslsWIcydu17EjzLCvIRNHZMJ3lfygtyQL7IAj1OLXXjN0EndZz3USfJPM8XKkrNFPw8o71U6sUWfV932Ur8ll-f8wEXPKQPTCDUCgyhYPqBChlvIVAXcaZsSYp5ZjFCVpUc9AaNIW5IVaOwCQAej2xAAD8img3JFtQlaMPldbJs275+g25h9sOo6UhWABOM6JibJa0NWzCNtuUFpFkbbhhe46UgSBIzt3LzyvEh1yBFV0dBzWs-HO37LvQtacPq+6ynBiBdpKA7jqSFEAHZvps55lrxzCCYIom+lSJ7IbemG6YpjEdBjKBJCwctw0QTE0DzdBEAARxANVEGUTTgBILEoBRbHbMZgGbpZ5Q2ZBSRQYhVIoVJ57ydegBWM6GkDCU7CAA))

<details> 
<summary> demo code for `for await` vs `for` </summary>

```ts
function resolveAfter(ms: number) {
  return new Promise<void>(resolve => { setTimeout(() => resolve(), ms) });
}

function * yield123Sync() {
    yield 1; yield 2; yield 3;
}

async function * yield123Async() { 
    await resolveAfter(100);
    yield 1; 
    await resolveAfter(100);
    yield 2; 
    await resolveAfter(100);
    yield 3;
}

function * yieldPromise123Sync() {
    yield resolveAfter(100).then(() => 1); 
    yield resolveAfter(100).then(() => 2); 
    yield resolveAfter(100).then(() => 3);
}

async function printValuesAsync<T>(iterable: Iterable<T> | AsyncIterable<T>) {
  const startTime = performance.now();
  console.log('starting to iterate (async flavor)');
  for await (const value of iterable) {
    console.log(value);
  }
  const endTime = performance.now();
  console.log(`done iterating (async flavor). Completed after ${endTime - startTime} ms`);

}


async function printValuesMixed<T>(iterable: Iterable<T> | AsyncIterable<T>) {
  const startTime = performance.now();
  console.log('starting to iterate (mixed flavor)');
  if (Symbol.asyncIterator in iterable){
    for await (const value of iterable) {
      console.log(value);
    }
  } else {
    for (const value of iterable) {
      console.log(value);
    }
  }
  const endTime = performance.now();
  console.log(`done iterating (mixed flavor). Completed after ${endTime - startTime} ms`);
}

async function main() {
  await printValuesAsync(yield123Sync()); // ~ 1 ms
  await printValuesAsync(yield123Async()); // ~309 ms
  await printValuesAsync(yieldPromise123Sync()); // ~311 ms

  console.log('compare to');

  await printValuesMixed(yield123Sync()); // ~2.7 ms
  await printValuesMixed(yield123Async()); // ~311 ms
  // the control flow here is quite different.
  await printValuesMixed(yieldPromise123Sync()); // ~5 ms
}

main()
```

</details>

***

So, anyway, what I'm really interested to understand, and don't currently understand, is - why would one end up with `AsyncIterable<T> | Iterable<T>` in the first place, and what control flow would a user of such an API _expect_ to happen if passing a possibly async-, possibly sync- iteratable to a Promise-returning function that takes either? Should it iterate their promises synchronously, then resolve, or await each promise sequentially? 

If it should be iterated synchronously, then I would say the API is better expressed as a non-promise-returning API in the first place, in which case the implementation will not be using `for await`.

If it should be iterated asynchronously/sequentially, then I would say that `for await` has incorrect error handling that permits floating promises, and one should consider instead some abomination like `await Promise.all(promises.map(async () => { /* do something after each promise */ });` or similar.

This is all to say, I'm certainly open to there being lenient handling of unions containing `AsyncIterable<T>`, I just really would like to have a good, real-world use to be able to point to as an example of why it's permitted if we do permit it (partially for my own interest! 🙂 ). I'm only marginally opposed by default in lieu of such an example.

**Comment by yume-chan** [comment]
My code uses `Iterable<T> | AsyncIterable<T>` for the same reason as `T | Promise<T>`: some implementations might be sync, the union type avoids unnecessary `async` keywords (and lint errors) on them.

From reading the above comments, what I understand is that the control flows of `Iterable<Promise<T>` and `AsyncIterable<T>` are different. But in my usage `T` is a concrete type (and I guess in generic usages, `T` would have a type constraint), so our functions will only accept `Iterable<T>` and `AsyncIterable<T>`, but not `Iterable<Promise<T>>`, thus the problem won't happen. 

So, the only difference between `for await ... of value` blindly and `Symbol.asyncIterator in value ? for await ... of value : for ... of value` is the extra microtasks, which I also don't care.

**Comment by lotmek** [comment]
Hello I just opened a PR to fix this issue,

However I don't know if that will resolve it for your example
```ts
const asyncIter: AsyncIterable<string> | Iterable<string> = [];
for await (const s of asyncIter) { }
```

I think Typescript will automatically deduct that `asyncIter` is only an array, and the variable won't be treated as a Union type...

**Comment by kirkwaiblinger** [comment]
> Hello I just opened a PR to fix this issue,
> 
> However I don't know if that will resolve it for your example
> 
> ```ts
> const asyncIter: AsyncIterable<string> | Iterable<string> = [];
> for await (const s of asyncIter) { }
> ```
> 
> I think Typescript will automatically deduct that `asyncIter` is only an array, and the variable won't be treated as a Union type...

Yeah, that's just TS being clever and narowing `asyncIter` based off its actual value, rather than treating it as the type in the type annotation. Technically we could write the repro more correctly as

```ts
declare const asyncIter: AsyncIterable<string> | Iterable<string>;

for await (const s of asyncIter) { }

export { }
```
or
```ts
async function foo(asyncIter: AsyncIterable<string> | Iterable<string>) {
  for await (const s of asyncIter) { }
}
```
